File: ruby/13-chatterbox/src/inventory_item.rb
# frozen_string_literal: true

class InventoryItem
  attr_accessor :id, :name, :batch_number, :cache_timestamp, :row_id, :quantity

  def initialize(id, name, batch_number, cache_timestamp, row_id, quantity)
    @id = id
    @name = name
    @batch_number = batch_number
    @cache_timestamp = cache_timestamp
    @row_id = row_id
    @quantity = quantity
  end

  def get_id
    @id
  end

  def set_id(v)
    @id = v
  end

  def get_name
    @name
  end

  def set_name(v)
    @name = v
  end

  def get_batch_number
    @batch_number
  end

  def set_batch_number(v)
    @batch_number = v
  end

  def get_cache_timestamp
    @cache_timestamp
  end

  def set_cache_timestamp(v)
    @cache_timestamp = v
  end

  def get_row_id
    @row_id
  end

  def set_row_id(v)
    @row_id = v
  end

  def get_quantity
    @quantity
  end

  def set_quantity(v)
    @quantity = v
  end

  def reserve(units)
    return { status: 'rejected', reason: 'invalid_quantity', remaining: @quantity } if units <= 0
    return { status: 'backorder', reserved: 0, remaining: @quantity } if units > @quantity

    @quantity -= units
    { status: 'reserved', reserved: units, remaining: @quantity, sku: "#{@id}-#{@batch_number}" }
  end

  def receive_stock(units)
    return @quantity if units <= 0

    @quantity += units
  end

  def public_snapshot
    {
      id: @id,
      name: @name,
      batch_number: @batch_number,
      quantity: @quantity,
      stock_status: stock_status
    }
  end

  def stock_status
    return 'out' if @quantity.zero?
    return 'low' if @quantity < 5

    'available'
  end
end

File: python/13-chatterbox/src/inventory_item.py
class InventoryItem:
    def __init__(self, id, name, batch_number, cache_timestamp, row_id, quantity):
        self.id = id
        self.name = name
        self.batch_number = batch_number
        self.cache_timestamp = cache_timestamp
        self.row_id = row_id
        self.quantity = quantity

    def get_id(self):
        return self.id

    def set_id(self, v):
        self.id = v

    def get_name(self):
        return self.name

    def set_name(self, v):
        self.name = v

    def get_batch_number(self):
        return self.batch_number

    def set_batch_number(self, v):
        self.batch_number = v

    def get_cache_timestamp(self):
        return self.cache_timestamp

    def set_cache_timestamp(self, v):
        self.cache_timestamp = v

    def get_row_id(self):
        return self.row_id

    def set_row_id(self, v):
        self.row_id = v

    def get_quantity(self):
        return self.quantity

    def set_quantity(self, v):
        self.quantity = v

    def reserve(self, units):
        if units <= 0:
            return {"status": "rejected", "reason": "invalid_quantity", "remaining": self.quantity}
        if units > self.quantity:
            return {"status": "backorder", "reserved": 0, "remaining": self.quantity}

        self.quantity -= units
        return {
            "status": "reserved",
            "reserved": units,
            "remaining": self.quantity,
            "sku": f"{self.id}-{self.batch_number}",
        }

    def receive_stock(self, units):
        if units <= 0:
            return self.quantity

        self.quantity += units
        return self.quantity

    def public_snapshot(self):
        return {
            "id": self.id,
            "name": self.name,
            "batch_number": self.batch_number,
            "quantity": self.quantity,
            "stock_status": self.stock_status(),
        }

    def stock_status(self):
        if self.quantity == 0:
            return "out"
        if self.quantity < 5:
            return "low"
        return "available"
